### DISCUSSION TITLE
Get the sha of the target branch on worklow run

### DISCUSSION THREAD

**Discussion by akoskm** [community/community#25367] [Actions]
<p>This is part of the script we used on GitLab to get the list of migration files we want to run:</p>
<pre><code class="lang-auto">  echo 'BRANCH_NAME='"$BRANCH_NAME"
  # The previous latest commit present on a branch before a merge request.
  # Only populated when there is a merge request associated with the pipeline. 
  echo 'CI_COMMIT_BEFORE_SHA='"$CI_COMMIT_BEFORE_SHA"
  # The commit revision for which project is built 
  echo 'CI_COMMIT_SHA='"$CI_COMMIT_SHA"

  MIGRATIONS=( $(git diff --name-only $CI_COMMIT_BEFORE_SHA $CI_COMMIT_SHA | grep -E 'migrations/[0-9]{4}-[0-9]{2}-[0-9]{2}-.*?\.sql') );
</code></pre>
<p>This generated us a list such as:</p>
<pre><code class="lang-auto">migrations/2020-05-30-add_deleted_at_to_constraint.sql
migrations/2020-06-04-add_location_id_constraint.sql
</code></pre>
<p>At the moment I can’t find a way to retrieve the list of files changes between the target branch and the source branch, during workflow run.</p>

**Comment by ghost** [comment]
<p>My colleague wrote a response in this thread, could you give their suggestion a try?</p><aside class="quote quote-modified" data-post="2" data-topic="17037">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/find-what-files-changed-in-a-pushed-commit/17037/2">Find what files changed in a pushed commit?</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    Please use below command to get the changed files in the push commit, please refer to this <a href="https://stackoverflow.com/questions/424071/how-to-list-all-the-files-in-a-commit" rel="nofollow noopener">link</a> for more details. 
git diff-tree --no-commit-id --name-only -r ${{ github.sha }}

[changedfiles.png]
  </blockquote>
</aside>

**Comment by akoskm** [comment]
<p>I tried this <a href="https://github.community/t/check-pushed-file-changes-with-git-diff-tree-in-github-actions/17220/2" class="inline-onebox">Check pushed file changes with git diff-tree in GitHub Actions</a> from the same thread because I need the list of changed files between multiple commits:</p>
<pre><code class="lang-auto">release commit (most recent commit)
migration-3 commit
migration-2 commit
new-feature-2 commit
bug-fix-1 commit
migration-1 commit
new-feature-1 commit
previous release commit (a commit from earlier)
</code></pre>
<p>So, in this case, I’ll need the 3 SQL files from migration-1, 2, and 3.</p>
<p>I’m obviously doing something wrong because I’m getting this message from the action:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0a5866a005abcbb7a5321b1ac7fd8bc12fc7f039.png" data-download-href="/uploads/short-url/1twaHULzU0kGD2N3plfF6jCWxHr.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181090625-daeaedc2-6010-4406-bc73-0b453bd998e9.png" alt="image" data-base62-sha1="1twaHULzU0kGD2N3plfF6jCWxHr" width="690" height="306" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0a5866a005abcbb7a5321b1ac7fd8bc12fc7f039_2_690x306.png, https://aws1.discourse-cdn.com/github/optimized/2X/0/0a5866a005abcbb7a5321b1ac7fd8bc12fc7f039_2_1035x459.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0a5866a005abcbb7a5321b1ac7fd8bc12fc7f039.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0a5866a005abcbb7a5321b1ac7fd8bc12fc7f039_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1067×474 44 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment]
<p>That log looks like you try to analyze the repository without first running <code>actions/checkout</code>, so the repository is not available. And since you want to look at the history, you might want to use its <a href="https://github.com/actions/checkout#Fetch-all-history-for-all-tags-and-branches" rel="nofollow noopener"><code>fetch-depth</code> parameter</a>, too.</p>

**Comment by akoskm** [comment]
<p>Thanks, I definitely miss checkout here. My current workflow looks like this:</p>
<pre><code class="lang-auto"># Github CI workflow
name: Github CI

on:
  push:
    paths-ignore:
    - '**/*.md'

env:
  CI_FLAG: "true"

jobs:
  build:
    runs-on: self-hosted
    container: my-container
    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 0

    - name: get changed files
      id: getfile
      run: |
        echo "::set-output name=files::$(git diff-tree --no-commit-id --name-only -r ${{ github.sha }} | xargs)"

    - name: echo output
      run: |
        echo ${{ steps.getfile.outputs.files }}
</code></pre>
<p>but the output is similar:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b764b764aef26b28d1add79f988cbc8b4ed66129.png" data-download-href="/uploads/short-url/qan9YEObr8LkkUYRkt4eOZzjwG5.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181090627-7042777b-1e18-4a18-a754-c8bfeb6c370c.png" alt="image" data-base62-sha1="qan9YEObr8LkkUYRkt4eOZzjwG5" width="690" height="213" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/b/b764b764aef26b28d1add79f988cbc8b4ed66129_2_690x213.png, https://aws1.discourse-cdn.com/github/optimized/2X/b/b764b764aef26b28d1add79f988cbc8b4ed66129_2_1035x319.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/b/b764b764aef26b28d1add79f988cbc8b4ed66129.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b764b764aef26b28d1add79f988cbc8b4ed66129_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1137×352 62.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p><strong>EDIT</strong><br>
I upgraded to git 2.27 on the instance that hosts github-runner.</p>

**Comment by akoskm** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> <a class="mention" href="/github-support">@github-support</a> I made some progress.<br>
Downgrading to checkout@v1 worked:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e522133f76f3f998d01599593637d718c4f4bd80.png" data-download-href="/uploads/short-url/wH0ltSYIBEMTvFsOdchaNjcCtHO.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181090631-3be2fcc5-183f-4092-a28b-97a2a1277758.png" alt="image" data-base62-sha1="wH0ltSYIBEMTvFsOdchaNjcCtHO" width="690" height="162" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/e/e522133f76f3f998d01599593637d718c4f4bd80_2_690x162.png, https://aws1.discourse-cdn.com/github/optimized/2X/e/e522133f76f3f998d01599593637d718c4f4bd80_2_1035x243.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/e/e522133f76f3f998d01599593637d718c4f4bd80.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e522133f76f3f998d01599593637d718c4f4bd80_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1047×247 27.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>What I’m missing here, why it doesn’t work with v2? :thinking:</p>
<p>I dug into the v2 repo. Is that possible that when I first run the runner my host had no git installed, checkout@v2 downloaded git 2.17, and even if I installed 2.27 later, checkout still works with the old git?</p>
<p>Switching back to v2 after v1 reports the same error like before.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/akoskm">@akoskm</a>,</p>
<p>All of your steps are executed in container ‘my-container’ on self-hosted runner. Is the git initailized in the container? Typically the error is due to not found the .git dir, or lack of permission.<br>
If you remove the container in the workflow yaml, can it be succesful with action v2?</p>
<p>Thanks.</p>

**Comment by akoskm** [comment]
<p><strong>Edit</strong></p>
<p>I ended up removing the runner, switching between checkout v1 and checkout v2 bricked my work folder. I also removed the container section.</p>
<p>After reinstalling the runner, checkout v1 was still thinking that I’m running v1, switching to v2 detected git 2.27 properly.</p>
<p>In the meantime, I deleted manually everything from my work folder <strong>_work</strong>.</p>
<p>Now I can run git commands during the workflow.</p>
<p><a class="mention" href="/airtower-luna">@airtower-luna</a>  unfortunately, checkout in combination with fetch-depth, still returns the files changed in the current commit:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/5e669c6dcfb5de95faf3377cf9f4ed75878bce74.png" data-download-href="/uploads/short-url/dt6IYnmxg9lfxUx3XWqeDe8Xdzu.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181090633-a3de5767-600a-441e-a3ad-70a17b0460b7.png" alt="image" data-base62-sha1="dt6IYnmxg9lfxUx3XWqeDe8Xdzu" width="690" height="274" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/5/5e669c6dcfb5de95faf3377cf9f4ed75878bce74_2_690x274.png, https://aws1.discourse-cdn.com/github/original/2X/5/5e669c6dcfb5de95faf3377cf9f4ed75878bce74.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/5/5e669c6dcfb5de95faf3377cf9f4ed75878bce74.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/5e669c6dcfb5de95faf3377cf9f4ed75878bce74_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1030×410 40 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>while the list of changed files between the target branch and the current commit is the following:</p>
<p><img src="https://user-images.githubusercontent.com/3111627/181090634-bc95f74e-d2d2-4223-b665-675c4602b42b.png" alt="image" data-base62-sha1="7DLbQD5fe6EfxfX0LuX4mql94wh" width="510" height="387"></p>
<p><strong>Original post below</strong></p>
<details>
<summary>
Summary</summary>
<blockquote>
<p>All of your steps are executed in container ‘my-container’ on self-hosted runner. Is the git initialized in the container?</p>
</blockquote>
<p>No, git (2.27) is initialized on the machine that hosts the self-hosted runner.</p>
<p>When I remove the container I get other errors:</p>
<pre><code class="lang-auto">Deleting the contents of '/home/ubuntu/actions-runner/_work/momentum/momentum'
##[error]Command failed: rm -rf "/home/ubuntu/actions-runner/_work/momentum/momentum/.github"
rm: cannot remove '/home/ubuntu/actions-runner/_work/momentum/momentum/.github/workflows/main.yml': Permission denied
rm: cannot remove '/home/ubuntu/actions-runner/_work/momentum/momentum/.github/ISSUE_TEMPLATE/migration.md': Permission denied
rm: cannot remove '/home/ubuntu/actions-runner/_work/momentum/momentum/.github/ISSUE_TEMPLATE/new-client.md': Permission denied
</code></pre>
<p>Would it help to uninstall the runner from the machine that hosts the self-hosted runner and reinstall it?</p>
</details>

**Comment by airtower-luna** [comment]
<p>That makes sense, when given one commit ID <code>diff-tree</code> shows the objects changed in that commit, you’ll have to use the version with two <code>&lt;tree-ish&gt;</code> parameters, see <a href="https://git-scm.com/docs/git-diff-tree" rel="nofollow noopener">documentation</a>.</p>
<p>The next problem will be to get the commit ID <em>before</em> the push. From looking at <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">context documentation</a> I think <code>github.event.before</code> might work, if I understand the <a href="https://developer.github.com/webhooks/event-payloads/#push" rel="nofollow noopener">push event payload description</a> correctly, but I haven’t tried that.</p>

**Comment by akoskm** [comment]
<p>Because I want to run the migrations only on branch <code>develop</code>, I realized I can get the hash of that branch with:</p>
<pre><code class="lang-auto">git rev-parse origin/develop
</code></pre>
<p>So the environment variable that was ready to use in GitLab CI <code>$CI_COMMIT_BEFORE_SHA</code>, but was missing from GitHub - if you hardcode it to develop is:</p>
<pre><code class="lang-auto">CI_COMMIT_BEFORE_SHA=$(git rev-parse origin/develop)
</code></pre>
<p><code>$CI_COMMIT_BEFORE_SHA</code> is available in GitHub runner as well and those are the only two hashes I needed for my old script.</p>
<p>Thanks <a class="mention" href="/github-support">@github-support</a>, <a class="mention" href="/airtower-luna">@airtower-luna</a>, <a class="mention" href="/weide-zhou">@weide-zhou</a>. :bowing_man:</p>

**Comment by akoskm** [comment]
<p>Disregard my previous answer, this doesn’t work.</p>
<p>The workflow runs <em>after</em> I merge the PR into development.</p>
<p><code>git rev-parse origin/develop</code> and <code>$CI_COMMIT_SHA</code> will be the same.</p>
<p>So again, is there any way to know what was SHA of the branch before the merge happened?</p>

**Comment by akoskm** [comment]
<p>Would it be possible to somehow use:</p>
<pre><code class="lang-auto">on:
  pull_request:
    types: [closed]
</code></pre>
<p>to retrieve the list of changed files in a PR. Then I could filter those and grab the ones ending with .sql.</p>
<p>This was I wouldn’t rely on SHAs.</p>
<p>FYI I tried the solution mentioned here <a href="https://github.community/t/get-list-of-files-on-pull-request-merge/17226/4" class="inline-onebox">Get list of files on pull request merge</a> and it didn’t work:</p>
<pre><code class="lang-auto">
Run URL="https://api.github.com/repos/${GITHUB_REPOSITORY}/pulls/76/files"
jq: error (at &lt;stdin&gt;:4): Cannot index string with string "filename"
##[error]Process completed with exit code 5.
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/akoskm">@akoskm</a>,</p>
<p>Please remove the jq code for your command firstly, make sure rest api(curl) command can get the pull request files content. Then follow the structure to add jq code to filter the filename.</p>
<p>Acccording to the jq error, there should be some code error in your script. I confirmed the sample code works fine on my side.</p>
<p>BTW, it’s recommened to check github context to get the sha…you wanted.</p>
<pre><code class="lang-auto">      - name: Dump GitHub context
        env:
          GITHUB_CONTEXT: ${{ toJson(github) }}
        run: echo "$GITHUB_CONTEXT"
</code></pre>
<p>Thanks.</p>

**Comment by akoskm** [comment, marked as answer]
<p>You’re right, I just made the following test:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/6534afdff72cf4c6b24674a6987ab6de5a01f895.png" data-download-href="/uploads/short-url/erj6z8NP3baZlo0arnV5cZfomG1.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181090637-e83a15ed-e903-45c5-8f22-4fdc65f2ccb2.png" alt="image" data-base62-sha1="erj6z8NP3baZlo0arnV5cZfomG1" width="690" height="130" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/6534afdff72cf4c6b24674a6987ab6de5a01f895_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">796×150 3.75 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>while in context I have:</p>
<pre><code class="lang-auto">"event": {
    "after": "85f578ec7e036f138b09c93ad1f252da4b7d10a3",
    "base_ref": null,
    "before": "4679d7e9916dcc9a1605d4b12a8b0a86f9c40f51",
</code></pre>
<p>This is clearly what I need, now I just have to find a way to get to <code>context.event.before/after</code> - I think it’s just <code>${{ github.event.before }}</code> - and send those SHA to my script.</p>
<p>Is there any difference these:</p>
<pre><code class="lang-auto">git diff --name-only ${{ github.event.before }} $GITHUB_SHA
git diff --name-only ${{ github.event.before }} HEAD
git diff --name-only ${{ github.event.before }} ${{ github.event.after }}
</code></pre>
<p>If I understand correctly all 3 would point to the latest commit in the branch.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/akoskm">@akoskm</a>,</p>
<p>The three commands could be different, it depends on which event you used.</p>
<p>If you use push event, <span class="math">GITHUB_SHA, HEAD, </span>{{ github.event.after }} all points to the latest commit, they have same value.</p>
<p>If you use pull request, they could be different:<br>
$GITHUB_SHA: Last merge commit on the  <code>GITHUB_REF</code>  branch, while GITHUB_REF is the fake merge branch  <code>refs/pull/:prNumber/merge</code>.</p>
<p>HEAD: same with GITHUB_SHA</p>
<p>${{ github.event.after }}: points to the latest commit on the compared(source) branch, not the fake merged branch.</p>
<p>Different sha will output different file list, hence, please make sure you used the correct sha for the git diff command.</p>
<p>Hope it clear and helpful!</p>

**Comment by akoskm** [comment]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a>,</p>
<p>is there are reason why context values set as environment variables are not being forwarded to my script? The syntax seems to be correct because <code>secrets</code> is visible in the script:</p>
<pre><code class="lang-auto">   - name: Run migrations
      shell: bash
      run: ./scripts/gitlab-migration.sh "$INT_DB_URL" "$CI_COMMIT_BEFORE_SHA"
      env:
       INT_DB_URL: ${{secrets.INT_DB_URL}}
       CI_COMMIT_BEFORE_SHA: ${{github.events.before}}
</code></pre>
<p>I can also see the values when I dump github context:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/bc41caaa8afa724db6646c8b7036663f00d602a0.png" data-download-href="/uploads/short-url/qRoIN2KO4CJ5Lny4vXDxFBWq66A.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181090640-f60d0a20-a21b-4ea3-9566-5de91367b376.png" alt="image" data-base62-sha1="qRoIN2KO4CJ5Lny4vXDxFBWq66A" width="384" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/b/bc41caaa8afa724db6646c8b7036663f00d602a0_2_384x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/b/bc41caaa8afa724db6646c8b7036663f00d602a0_2_576x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/b/bc41caaa8afa724db6646c8b7036663f00d602a0_2_768x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/bc41caaa8afa724db6646c8b7036663f00d602a0_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">878×1142 74.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/akoskm">@akoskm</a>,</p>
<p>Please use ${{ github.event.before }}, no s for event.</p>

**Comment by akoskm** [comment]
<p>Thanks <a class="mention" href="/weide-zhou">@weide-zhou</a>  for catching that typo. :sweat_smile:</p>
<p>Works as expected:</p>
<p><img src="https://user-images.githubusercontent.com/3111627/181090643-886ac763-e08a-4f24-83b1-db0e93b5f87e.png" alt="image" data-base62-sha1="l8YeIBsUrbaAgNbhgTlPjlcT6aI" width="612" height="223"></p>
